tsl/src/bgw_policy/continuous_aggregate_api.c

```diff
@@ -39,3 +54,236 @@ policy_continuous_aggregate_proc(PG_FUNCTION_ARGS)
+	Hypertable *mat_ht;
+	Dimension *dim;
+	ContinuousAgg *cagg;
+	int32 job_id, mat_htid;
+	Datum start_interval, end_interval;
+	Interval refresh_interval;
+	Oid dim_type, start_interval_type, end_interval_type;
+	Oid cagg_oid, owner_id;
+	List *jobs;
+	bool if_not_exists;
+	if (PG_ARGISNULL(1) || PG_ARGISNULL(2) || PG_ARGISNULL(3))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot add refresh policy due to null argument values"),
+				 errhint("Please provide non-null values for start_interval, end_interval and "
+						 "schedule_interval.")));
+	}
+	cagg_oid = PG_GETARG_OID(0);
+	start_interval = PG_GETARG_DATUM(1);
+	end_interval = PG_GETARG_DATUM(2);
+	start_interval_type = get_fn_expr_argtype(fcinfo->flinfo, 1);
+	end_interval_type = get_fn_expr_argtype(fcinfo->flinfo, 2);
+	refresh_interval = *PG_GETARG_INTERVAL_P(3);
+	if_not_exists = PG_GETARG_BOOL(4);
+	cagg = ts_continuous_agg_find_by_relid(cagg_oid);
+	if (!cagg)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("could not add refresh policy because \"%s\" is not a continuous aggregate",
+						get_rel_name(cagg_oid))));
+	}
+
+	mat_htid = cagg->data.mat_hypertable_id;
+	/* Verify that the owner can create a background worker */
+	owner_id = ts_cagg_permissions_check(cagg_oid, GetUserId());
+	hcache = ts_hypertable_cache_pin();
+	mat_ht = ts_hypertable_cache_get_entry_by_id(hcache, mat_htid);
+	dim = hyperspace_get_open_dimension(mat_ht->space, 0);
+	dim_type = ts_dimension_get_partition_type(dim);
+	ts_cache_release(hcache);
+	ts_bgw_job_validate_job_owner(owner_id, JOB_TYPE_CONTINUOUS_AGGREGATE);
+
+	/* Make sure there is only 1 refresh policy on the cagg */
+	jobs = ts_bgw_job_find_by_proc_and_hypertable_id(POLICY_REFRESH_CAGG_PROC_NAME,
+													 INTERNAL_SCHEMA_NAME,
+													 mat_htid);
+
+	if (jobs != NIL)
+	{
+		Assert(list_length(jobs) == 1);
+		if (!if_not_exists)
+			ereport(ERROR,
+					(errcode(ERRCODE_DUPLICATE_OBJECT),
+					 errmsg("refresh policy already exists for continuous aggregate \"%s\"",
+							get_rel_name(cagg_oid))));
+		Assert(list_length(jobs) == 1);
+		BgwJob *existing = linitial(jobs);
+
+		if (policy_config_check_ht_lag_equality(existing->fd.config,
+												CONFIG_KEY_START_INTERVAL,
+												dim_type,
+												start_interval_type,
+												start_interval) &&
+			policy_config_check_ht_lag_equality(existing->fd.config,
+												CONFIG_KEY_END_INTERVAL,
+												dim_type,
+												end_interval_type,
+												end_interval))
+		{
+			/* If all arguments are the same, do nothing */
+			ereport(NOTICE,
+					(errmsg("refresh policy already exists on continuous aggregate \"%s\", "
+							"skipping",
+							get_rel_name(cagg_oid))));
+			PG_RETURN_INT32(-1);
+		}
+		else
+		{
+			elog(WARNING,
+				 "could not add refresh policy due to existing policy on continuous aggregate with "
+				 "different arguments");
+			PG_RETURN_INT32(-1);
```

Comment:
Should this be an error? If not, why so?

```suggestion
			elog(WARNING,
				 "an incompatible policy already exists for continuous aggregate \"%s\"", get_rel_name(cagg_oid));
			PG_RETURN_INT32(-1);
```

Reply:
This is consistent with the other policy apis (error based on  if if_not_exists setting). Also think the previous error message is more explicit and also consistent with the error messages for the other polices.